File: src/0989.add-to-array-form-of-integer.1031/solution.ts
function addToArrayForm (A: number[], K: number): number[] {
    A.reverse();
    let addon = 0;
    let index = 0;
    while (index < A.length && K > 0) {
        const kDigit = K % 10;
        K = (K - kDigit) / 10;
        const sum = addon + A[index] + kDigit;
        const digit = sum % 10;
        A[index++] = digit;
        addon = (sum - digit) / 10;
    }
    while (addon && index < A.length) {
        const sum = addon + A[index];
        const digit = sum % 10;
        A[index++] = digit;
        addon = (sum - digit) / 10;
    }

    while (K > 0) {
        const kDigit = K % 10;
        K = (K - kDigit) / 10;
        const sum = addon + kDigit;
        const digit = sum % 10;
        A.push(digit);
        addon = (sum - digit) / 10;
    }

    if (addon) {
        A.push(addon);
    }

    return A.reverse();
}

File: src/0989.add-to-array-form-of-integer.1031/solution.java
class Solution {
    public List<Integer> addToArrayForm(int[] num, int k) {
        List<Integer> result = new ArrayList<>();
        int rest = 0;
        int sum = 0;
        int index = num.length-1;
        while(index>-1 && k>0){
            sum = rest+num[index--]+(k%10);
            result.add(sum%10);
            rest = sum/10;
            k /= 10;
        }
        
        while(index>-1){
            sum = rest + num[index--];
            result.add(sum%10);
            rest = sum/10;
        }
        
        while (k>0){
            sum = rest+(k%10);
            result.add(sum%10);
            rest = sum/10;
            k /= 10;
        }
        
        if(rest != 0){
            result.add(rest);
        }

        return result.reversed();
    }
}